### PR TITLE
texture atlasing to reduce file sizes

### PR DESCRIPTION
**Pull request by fin444**
`/media/` size reduced by 102.6kb, though we won't necessarily see the exact same reduction in page loads

### CODE REVIEW

File: js/icons.js [hunk 1]
```diff
@@ -1,33 +1,39 @@
 //which icons to use
 let window_storage = window.localStorage;
+console.log(window_storage);
```
Comment (hppeng-wynn) [inline js/icons.js:3]: minor: remove

File: js/icons.js [hunk 2]
```diff
@@ -1,33 +1,39 @@
...
 icon_state_stored = window_storage.getItem("newicons");
 newIcons = true;
 if (icon_state_stored === "false") {toggleIcons()} 
 
+
 /** Toggle icons on the ENTIRE page.
  * 
  */
-function toggleIcons() {
+ function toggleIcons() {
```
Comment (hppeng-wynn) [inline js/icons.js:12]: style: remove leading space

File: js/icons.js [hunk 3]
```diff
@@ -1,33 +1,39 @@
...
         favicon.href = favicon.href.replace("media/icons/old","media/icons/new");
         for (const img of imgs) {
             if (img.src.includes("media/icons/old")) {img.src = img.src.replace("media/icons/old","media/icons/new");}
-            if (img.src.includes("media/items/old")) {img.src = img.src.replace("media/items/old","media/items/new");}
         }
-        toggleiconbutton.textContent = "Use Old Icons";
+        for (const div of divs) {
+            if (div.style.backgroundImage.includes("media/items/old")) {div.style.backgroundImage = div.style.backgroundImage.replace("media/items/old","media/items/new");}
+        }
+        //toggleiconbutton.textContent = "Use Old Icons";
```
Comment (hppeng-wynn) [inline js/icons.js:26]: maybe label all the atlas images with a class or something? to make this a bit cleaner
https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName

though idk how this performs, if it does a linear scan internally or if there's something more optimized / internal data structures to speed it up

though it should be more optimized in any case (linear scan in native vs js)

File: js/icons.js [hunk 4]
```diff
@@ -1,33 +1,39 @@
...
     } else { //switch to old
         favicon.href = favicon.href.replace("media/icons/new","media/icons/old");
         for (const img of imgs) {
             if (img.src.includes("media/icons/new")) {img.src = img.src.replace("media/icons/new","media/icons/old");}
-            if (img.src.includes("media/items/new")) {img.src = img.src.replace("media/items/new","media/items/old");}
         }
-        toggleiconbutton.textContent = "Use New Icons";
+        for (const div of divs) {
+            if (div.style.backgroundImage.includes("media/items/new")) {div.style.backgroundImage = div.style.backgroundImage.replace("media/items/new","media/items/old");}
+        }
```
Comment (hppeng-wynn) [inline js/icons.js:35]: i know this isn't your code per se, but this if statement should be merged to reduce duplicate code.

```
let resource_base;
if (...) {resource_base = ...}
else {resource_base = ...}

...
```

File: js/atree.js [hunk 5]
```diff
@@ -1280,6 +1250,62 @@ function atree_set_state(node_wrapper, new_state) {
...
+    "node_1": 1,
+    "node_2": 2,
+    "node_3": 3,
+    "node_archer": 4,
+    "node_warrior": 5,
+    "node_mage": 6,
+    "node_assassin": 7,
+    "node_shaman": 8
+}
+const atreeNodeAtlasSize = [17, 5]
```
Comment (hppeng-wynn) [inline js/atree.js:1280]: i don't think? the node atlas needs to be 17x5 -- TODO: do some math to compute good sizes

i think it had good (integer/half-integer) %s out of the box with (9, 3) (probably goes down to (8, 2) which is good..? i haven't calculated it)

File: js/atree.js [hunk 6]
```diff
@@ -1280,6 +1250,62 @@ function atree_set_state(node_wrapper, new_state) {
...
+        let connector_info = atree_connectors_map.get(i);
+        let connector_elem = connector_info.connector;
+        set_connector_type(connector_info);
+        connector_info.highlight = [0, 0, 0, 0];
+        connector_elem.style.backgroundPosition = atlasBGPositionCalc(atreeConnectorAtlasPositions[connector_info.type]["0000"], atreeConnectorAtlasSize);
+        let target_elem = document.getElementById("atree-row-" + i.split(",")[0]).children[i.split(",")[1]];
+        if (target_elem.children.length != 0) {
+            // janky special case...
+            connector_elem.style.display = 'none';
+        }
```
Comment (hppeng-wynn) [inline js/atree.js:1303]: oh god

i was gonna suggest changing this but this is disgusting and necessary

nvm!

File: builder/index_full.html [hunk 7]
```diff
@@ -49,7 +49,7 @@
                             <div class="col-auto rounded order-xl-0 order-0">
                                 <div class="row h-100 dark-shadow dark-6 rounded" id="helmet-dropdown">
                                     <div class="col-auto px-lg-1 g-0 dark-7 rounded-end my-auto text-center scaled-item-icon" id="helmet-img-loc">
-                                        <img id="helmet-img" class="img-fluid rounded" src="../media/items/new/generic-helmet.png">
+                                        <div id="helmet-img" class="img-fluid rounded" style="background-image: url('../media/items/new.png'); image-rendering: pixelated; background-size: 1200% 100%; background-position: 45.45454545454546% 0; aspect-ratio: 1/1;"></div>
```
Comment (hppeng-wynn) [inline builder/index_full.html:52]: might help reduce code duplication if a css class is made for the "equipment atlas" images, and then inline style is used only for positioning. Though that might be more confusing for future code readers

**Review by hppeng-wynn — CHANGES_REQUESTED** [review]
neat neat

a bit busy rn unfortunately but ill test some more soonTM, only looked at code atm

File: js/atree.js (hunk 6 above)
Comment (fin444) [reply js/atree.js:1303]: I didn't write this code, I just moved it.

Comment (hppeng-wynn) [reply js/atree.js:1303]: yep
and i remember my own stupid, stupid display code

File: js/icons.js [hunk 8]
```diff
@@ -1,33 +1,27 @@
...
+    let imgs = document.getElementsByTagName("img");
+    let divs = document.getElementsByClassName("item-display-new-toggleable");
+    let favicon = document.querySelector("link[rel~='icon']");
+    favicon.href = favicon.href.replace("media/icons/" + (newIcons ? "old" : "new"), "media/icons/" + newOrOld);
+    for (const img of imgs) {
+        // if doesn't contain, replace() does nothing
+        img.src = img.src.replace("media/icons/" + (newIcons ? "old" : "new"), "media/icons/" + newOrOld);
+    }
+    for (let i = 0; i < divs.length; i++) {
+        console.log(divs.item(i))
```
Comment (hppeng-wynn) [inline js/icons.js:24]: remove console log

**Review by hppeng-wynn — CHANGES_REQUESTED** [review]
just one thing

**Review by hppeng-wynn — APPROVED** [review]
nice

**Pull request merged**